### PR TITLE
services.openssh: add banner item

### PR DESCRIPTION
**Pull request by teto**
Add the possibility to configure a banner.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

### CODE REVIEW

**Comment by teto** [comment]
Not sure who to ping, so @Infinisil pardon me :p

As a general remark, it's a bit annoying that such a critical module doesn't have any maintainers listed.

File: nixos/modules/services/networking/ssh/sshd.nix
Comment (Infinisil) [inline nixos/modules/services/networking/ssh/sshd.nix]: I think `nullOr lines` with `default = null` would be better so people can set empty banners too.

File: nixos/modules/services/networking/ssh/sshd.nix
Comment (Infinisil) [inline nixos/modules/services/networking/ssh/sshd.nix]: Oh how I wish I could put structural `settings` options everywhere as described in NixOS/rfcs#42. I think it's not a problem here because openssh handles multiple assignments gracefully, but if it didn't this could break for people who set `Banner` in `extraConfig` themselves as described in https://github.com/NixOS/rfcs/pull/42#issuecomment-523137258

File: nixos/modules/services/networking/ssh/sshd.nix
Comment (teto) [reply nixos/modules/services/networking/ssh/sshd.nix]: Looking at sshd_config: ```
       Banner The contents of the specified file are sent to the remote user before authentication is allowed.  If the argument is none then no banner is displayed.  By default, no banner
              is displayed.
```
So I would say we should instead always write "Banner" to the configuration and just default it to "none" instead ?

Comment (Infinisil) [reply nixos/modules/services/networking/ssh/sshd.nix]: I see, yeah using `types.lines` with a default of `"none"` sounds reasonable

**Comment by teto** [comment]
using "none" as a default was not to my taste in the end (also it prevented having a banner with the content "none") so I made null the default as you initally suggested. My next concern is that Banner expecting a path, maybe it would be best to just accept a path instead of lines ? (though I think in 90% cases, lines is more convenient).

File: nixos/modules/services/networking/ssh/sshd.nix
Comment (Infinisil) [inline nixos/modules/services/networking/ssh/sshd.nix:463]:
```suggestion
        Banner ${if cfg.banner == null then "none" else pkgs.writeText "ssh_banner" cfg.banner}
```

**Review by infinisil — CHANGES_REQUESTED** [review]
Looks good to merge other than the nit below

**Review by infinisil — APPROVED** [review]

**Comment by Infinisil** [comment]
Can you squash the commits into one? Will merge after that :)

**Comment by teto** [comment]
@Infinisil what's the difference with squashing from GH's UI ?

**Comment by Infinisil** [comment]
- It maintains the original commit sha1 and signature if there is one
- Squash button creates a single commit, which isn't always ideal
- With the squash button I have to combine commit messages, even though these aren't any of my commits

**Comment by teto** [comment]
thank you for the explanation. Squashed :)

**Pull request merged**
